### PR TITLE
Create ObjectStore from URL and Options (#4047)

### LINKED ISSUE #4047: object_store: Instantiate object store from provided url with store options
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Currently, in the projects that are using `object_store` - datafusion/ delta-rs/ pola-rs etc, a `dyn ObjectStore` has to be created manually by parsing the provided URL, checking the scheme and providing the options.

It would be great to have this capability directly provided by the crate.

**Describe the solution you'd like**
My proposal is to standardize this implementation and bring it into this crate itself exposed by a simple function like the below as it would make things significantly simple for developers using the crate.

```
/// Taken a reference from implementation in delta-rs 
https://github.com/delta-io/delta-rs/blob/c8371b38fdf22802f0f91b4ddc2a47da6be97c68/rust/src/storage/mod.rs#L85-L100

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct StorageOptions(pub HashMap<String, String>);

#[derive(Debug, Clone)]
pub struct AObjectStore {
    storage: Arc<dyn ObjectStore>,
    location: Url,
    options: StorageOptions,
}

/// Try creating a new instance of [`AObjectStore`]
pub fn get_object_store(location: Url, options: impl Into<StorageOptions> + Clone) -> Result<AObjectStore> {
    let prefix = Path::from(location.path());
   
    // parse URL to a kind (s3/ aws/ ... )
    let kind = ObjectStoreKind::parse_url(&location)?;

    // instantiate object store
    let store = kind.into_impl( .... );

    // return
    Ok(Self {
        store,
        location,
        options: options.into(),
    })
}
```

For any new storage backends that may come up in the future, they can be added to the `ObjectStoreKind` along with a small implementation in `into_impl`. Users of the crate will only have to bump up the crate version. 

**Describe alternatives you've considered**
Without this, each lib using `object_store` has to implement its own parsing.

Examples: 
- See datafusion registry [here](https://github.com/apache/arrow-datafusion/blob/52fa2285b43ad6712e9b8bf6c05b4b8ff93f44f9/datafusion/execution/src/object_store.rs#L186-L217).
- See delta-rs implementation [here](https://github.com/delta-io/delta-rs/blob/c8371b38fdf22802f0f91b4ddc2a47da6be97c68/rust/src/storage/config.rs#L138-L196)

Also without this each time this crate adds a new backend, the users of this crate will have to bump up the version and add implementation for the backends by themselves. 

**Additional context**
This idea is also implemented by,

- PyArrow FileSystem API [fs.FileSystem.from_uri(uri)](https://arrow.apache.org/docs/python/generated/pyarrow.fs.FileSystem.html#pyarrow.fs.FileSystem.from_uri)
- Hadoop FileSystem API [org.apache.hadoop.fs.FileSystem.get(uri, conf)](https://hadoop.apache.org/docs/r3.0.0/api/org/apache/hadoop/fs/FileSystem.html#get-java.net.URI-org.apache.hadoop.conf.Configuration-)
- Fsspec [fsspec.filesystem(protocol, **storage_options)](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.filesystem)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4047

# Rationale for this change
 

See ticket

# What changes are included in this PR?

Adds `parse_url` and `parse_url_opts` 

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/parse.rs [hunk 1]
```diff
@@ -100,40 +114,41 @@ macro_rules! builder_opts {
...
 }
 
-/// Returns a [`ObjectStore`] for the provided `url`
-pub fn parse_url(url: &Url) -> Result<Box<dyn ObjectStore>, super::Error> {
+/// Create an [`ObjectStore`] based on the provided `url`
+///
+/// Returns
+/// - An [`ObjectStore`] of the corresponding type
+/// - The [`Path`] into the [`ObjectStore`] of the addressed resource
+pub fn parse_url(url: &Url) -> Result<(Box<dyn ObjectStore>, Path), super::Error> {
```
Comment (tustvold) [inline object_store/src/parse.rs:126]: This signature is consistent with pyarrow [from_uri](https://arrow.apache.org/docs/python/generated/pyarrow.fs.FileSystem.html#pyarrow.fs.FileSystem.from_uri).

It allows downstreams to determine whether they want the remaining path to be used to "namespace" the store, i.e. using `PrefixStore` (delta-rs), or just want the corresponding `ObjectStore` for the path (datafusion)

File: object_store/src/aws/mod.rs [hunk 2]
```diff
@@ -721,7 +721,7 @@ impl AmazonS3Builder {
     ///
     /// - `s3://<bucket>/<path>`
     /// - `s3a://<bucket>/<path>`
-    /// - `https://s3.<bucket>.amazonaws.com`
+    /// - `https://s3.<region>.amazonaws.com/<bucket>`
```
Comment (tustvold) [inline object_store/src/aws/mod.rs:724]: This is a holdover from #4082

**Comment by tustvold** [comment]
@roeap and @chitralverma I think I am finally happy with this, PTAL

File: object_store/src/parse.rs [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+            _ => Err(Error::Unrecognised { url: url.clone() }),
+        }
+    }
+}
+
+#[cfg(any(feature = "aws", feature = "gcp", feature = "azure", feature = "http"))]
+macro_rules! builder_opts {
+    ($builder:ty, $url:expr, $options:expr) => {{
+        let builder = $options.into_iter().fold(
+            <$builder>::from_env().with_url($url.as_str()),
```
[the remark this answers is not in the document]
Comment (roeap) [reply object_store/src/parse.rs:111]: This is actually something we have been struggling a bit with, and I am unsure what the best solution is without introducing too much complexity.

Essentially we end up in a situation where any valid credentials configured in the environment could take precedence over a credentials passed via the storage options. Priority would be based on the order in which we check credentials during build.

To work around this I was contemplating several options, none of which really resonated so far. but they all boil down to having some sort of notion on what combinations of config options constitute a credential, then checking if the options contain a full credential and skipping parsing from env if they do.

Things get a bit messier if the options contain a partial credential that can be imputed via the environment, as especially in the azure case there are several permutations. Within delta-rs we are unfortunately in a situation right now, where this is affecting users.

This becomes even more relevant as we begin to introduce catalog integrations where different locations are associated with different credentials.

[the remark this answers is not in the document]
Comment (tustvold) [reply object_store/src/parse.rs:111]: Perhaps we just don't call from_env and leave downstreams to populate keys from the environment if they so wish?

[the remark this answers is not in the document]
Comment (roeap) [reply object_store/src/parse.rs:111]: Sounds good!

I can iterate on a solution in delta-rs and if we find something feasible bring this upstream?

**Review by roeap — APPROVED** [review]
very nice - looking forward to integrating this!

**Pull request merged**